Support `CharSequence` argument for Fallback String-to-Object Conversion

sbrannen opened the pull request:
Prior to this commit, the ["Fallback String-to-Object Conversion"](https://docs.junit.org/current/user-guide/#writing-tests-parameterized-tests-argument-conversion-implicit-fallback) support for parameterized tests supported factory constructors and methods that accepted a single `String` argument.

This commit relaxes that restriction to support factory constructors and methods that accept either a single `String` argument or a single `CharSequence` argument, since the original source `String` can always be supplied to an `Executable` that accepts a `CharSequence`.

Note that this change is available to third parties via `ConversionSupport` in `junit-platform-commons`, which `junit-jupiter-params` utilizes.

- Closes #4815

sbrannen commented:
In my opinion, both the feature and the implementation are straightforward.

However, after having put a little more thought into it, I believe we may want to alter the implementation to support backward compatibility.

For example, if an existing code base has overloaded `from(String)` and `from(CharSequence)` factory methods, JUnit would have used `from(String)` prior to 6.0. However, with this change in 6.0, JUnit will ignore both `from(...)` factory methods, since we only honor a single, unambiguous factory method.

Thus, we should ideally make a first pass to find `String` factory methods/constructors to support backward compatibility and then make additional pass to find `CharSequence` factories.

In other words, I think this new feature should effectively be a "fallback for the fallback". 😉 

I'm happy to implement that, but I wanted to bring up the topic for discussion first.

sbrannen commented on junit-platform-commons/src/main/java/org/junit/platform/commons/support/conversion/FallbackStringToObjectConverter.java:
```diff
@@ -160,15 +160,21 @@ public boolean test(Constructor<?> constructor) {
+			return false;
+		}
+		Object parameterType = executable.getParameterTypes()[0];
+		return (parameterType == String.class || parameterType == CharSequence.class);
```
No. The source will always be a `String`.

Thus, we can only support something that a `String` can be cast to, which includes `CharSequence` and `Object`.

I originally had something like `parameterType.isAssignableFrom(String.class)`, but that was too broad, because we don't actually want to invoke a factory method like `from(Object)`.

And actually, technically... a `String` is an 'instance of' `Object`, `java.io.Serializable`, `Comparable<String>`, `CharSequence`, `Constable`, and `ConstantDesc`. So, we'd better really limit the check to what we intend to support. 😇

sormuras replied:
`String` is `final`. No need to check for subtypes.

> And actually, technically... a String is an 'instance of' Object, java.io.Serializable, Comparable<String>, CharSequence, Constable, and ConstantDesc. So, we'd better really limit the check to what we intend to support. 😇

But why? The first two checks are fast, only the third would cost more.

return (
     parameterType == String.class
  || parameterType == CharSequence.class
  || parameterType.isAssignableFrom(CharSequence.class)
  );

sbrannen replied:
To ensure we don't mess this up, I introduced the following assertion in the tests and force pushed.

```java
assertThat(isBookFactoryMethod).rejects(bookMethod("factory", StringBuilder.class));
```

sbrannen commented:
> For example, if an existing code base has overloaded `from(String)` and `from(CharSequence)` factory methods, JUnit would have used `from(String)` prior to 6.0. However, with this change in 6.0, JUnit will ignore both `from(...)` factory methods, since we only honor a single, unambiguous factory method.

I've marked this PR as a _draft_ until I've reworked the implementation to avoid introducing that regression.

sbrannen commented:
> > For example, if an existing code base has overloaded `from(String)` and `from(CharSequence)` factory methods, JUnit would have used `from(String)` prior to 6.0. However, with this change in 6.0, JUnit will ignore both `from(...)` factory methods, since we only honor a single, unambiguous factory method.
> 
> I've marked this PR as a _draft_ until I've reworked the implementation to avoid introducing that regression.

I have addressed this, force pushed, and removed draft status.
